test/integration/scheduler/predicates_test.go

```diff
@@ -1219,25 +1235,251 @@ func TestPodTopologySpreadFilter(t *testing.T) {
+				st.MakeNode().Name("node-3").Label("node", "node-3").Label("zone", "zone-2").Label("foo", "").Obj(),
+				st.MakeNode().Name("node-4").Label("node", "node-4").Label("zone", "zone-2").Obj(),
+			},
+			candidateNodes:             []string{"node-1", "node-3"},
+			enableNodeInclustionPolicy: true,
+		},
+		{
+			// 1. to fulfil "zone" constraint, pods spread across zones as ~3~/~1~
+			// 2. to fulfil "node" constraint, pods spread across zones as 1/~0~/0/~0~
+			// node-2 and node 4 are filtered out by plugins
+			name: "soft constraint with two Constraints on node and zone, one ignored NodeAffinityPolicy, one honored NodeTaintsPolicy",
+			incomingPod: st.MakePod().Name("p").Label("foo", "").Container(pause).
+				NodeSelector(map[string]string{"foo": ""}).
+				SpreadConstraint(1, "zone", v1.ScheduleAnyway, st.MakeLabelSelector().Exists("foo").Obj(), nil, &ignorePolicy, nil).
+				SpreadConstraint(1, "node", v1.ScheduleAnyway, st.MakeLabelSelector().Exists("foo").Obj(), nil, nil, &honorPolicy).
+				Obj(),
+			existingPods: []*v1.Pod{
+				st.MakePod().Name("p1a").Node("node-1").Label("foo", "").Container(pause).Obj(),
+				st.MakePod().Name("p2a").Node("node-2").Label("foo", "").Container(pause).Obj(),
+				st.MakePod().Name("p2b").Node("node-2").Label("foo", "").Container(pause).Obj(),
+				st.MakePod().Name("p4a").Node("node-4").Label("foo", "").Container(pause).Obj(),
+			},
+			fits: true,
+			specificNodes: []*v1.Node{
+				st.MakeNode().Name("node-1").Label("node", "node-1").Label("zone", "zone-1").Label("foo", "").Obj(),
+				st.MakeNode().Name("node-2").Label("node", "node-2").Label("zone", "zone-1").Label("foo", "").Taints(taints).Obj(),
+				st.MakeNode().Name("node-3").Label("node", "node-3").Label("zone", "zone-2").Label("foo", "").Obj(),
+				st.MakeNode().Name("node-4").Label("node", "node-4").Label("zone", "zone-2").Obj(),
+			},
+			candidateNodes:             []string{"node-1", "node-3"},
+			enableNodeInclustionPolicy: true,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.MinDomainsInPodTopologySpread, tt.enableMinDomains)()
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.PodTopologySpreadNodePolicy, tt.enableNodeInclustionPolicy)()
+
 			testCtx := initTest(t, "pts-predicate")
 			cs := testCtx.ClientSet
 			ns := testCtx.NS.Name
 			defer testutils.CleanupTest(t, testCtx)
 
-			for i := 0; i < 4; i++ {
-				// Create nodes with labels "zone: zone-{0,1}" and "node: <node name>" to each node.
-				nodeName := fmt.Sprintf("node-%d", i)
-				zone := fmt.Sprintf("zone-%d", i/2)
-				_, err := createNode(cs, st.MakeNode().Name(nodeName).Label("node", nodeName).Label("zone", zone).Obj())
-				if err != nil {
-					t.Fatalf("Cannot create node: %v", err)
+			if len(tt.specificNodes) != 0 {
```

Comment:
why introduce this field? I noticed the added cases are all 4 nodes, are labelled the same.

Reply:
Labels are not exactly the same, some nodes hold affinity labels matching selector in `constraint`, what's more, some nodes should have taints to verify `NodeTaintsPolicy`.

Reply:
rename the field to `nodes`

Then, create a variable `defaultNodes` and add it to the test cases that where expecting them.